File: libs/power-log-parser/src/lib/parsers/tourist-revealed-parser.ts
import { BlockType, CardIds, GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, FullEntity, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class TouristRevealedParser implements ActionParser {
	readonly ParserName = 'TouristRevealedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.Action &&
			(node.Object as Action).Type === (BlockType.TRIGGER as number) &&
			this.GameState.CurrentEntities.get((node.Object as Action).Entity)?.CardId ===
				CardIds.TouristVfxEnchantment_VAC_422e
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		let touristEntity: FullEntity | null =
			action?.Data.filter((d): d is FullEntity => d instanceof FullEntity).find(
				(e) => e.GetTag(GameTag.TOURIST) > 0,
			) ?? null;

		if (touristEntity == null) {
			const parent = node.Parent;
			if (parent?.Type === NodeType.Action) {
				const parentAction = parent.Object as Action;
				const maybeTourist = this.GameState.CurrentEntities.get(parentAction.Entity);
				if (maybeTourist && maybeTourist.GetTag(GameTag.TOURIST) > 0) {
					touristEntity = maybeTourist;
				}
			}
		}

		if (touristEntity == null) {
			return null;
		}

		return [
			GameEventProvider.Create(
				action.TimeStamp,
				'TOURIST_REVEALED',
				GameEventHelper.CreateProvider(
					'TOURIST_REVEALED',
					touristEntity.CardId,
					touristEntity.GetController(),
					touristEntity.Id,
					this.StateFacade,
					{
						TouristFor: touristEntity.GetTag(GameTag.TOURIST),
					},
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/TouristRevealedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData.Meta;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using System.Linq;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events.Parsers
{
    public class TouristRevealedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public TouristRevealedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(Action)
                && (node.Object as Action).Type == (int)BlockType.TRIGGER
                && GameState.CurrentEntities.GetValueOrDefault((node.Object as Action).Entity)?.CardId == CardIds.TouristVfxEnchantment_VAC_422e;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Action;
            var touristEntity = action?.Data
                .Where(d => d is FullEntity)
                .Select(d => d as FullEntity)
                .Where(e => e.GetTag(GameTag.TOURIST) > 0)
                .FirstOrDefault();
            if (touristEntity == null)
            {
                var parent = node.Parent;
                if (parent.Type == typeof(Action))
                {
                    var parentAction = node.Parent.Object as Action;
                    var maybeTourist = GameState.CurrentEntities.GetValueOrDefault(parentAction.Entity);
                    if (maybeTourist.GetTag(GameTag.TOURIST) > 0)
                    {
                        touristEntity = maybeTourist;
                    }
                }
            }

            if (touristEntity == null)
            {
                return null;
            }

            List<GameEventProvider> result = new List<GameEventProvider>();
            result.Add(GameEventProvider.Create(
                action.TimeStamp,
                "TOURIST_REVEALED",
                GameEvent.CreateProvider(
                    "TOURIST_REVEALED",
                    touristEntity.CardId,
                    touristEntity.GetController(),
                    touristEntity.Id,
                    StateFacade,
                    //null,
                    new
                    {
                        TouristFor = touristEntity.GetTag(GameTag.TOURIST)
                    }),
                true,
                node));
            return result;
        }
    }
}
